hashintel/hash
# Support multiple data types for arrays on properties table

## 🌟 What is the purpose of this PR?

This PR adds support for multiple data types on array/list properties at entity editor. So users can select which data type they want to add to the list.

## 🔗 Related links

- [Asana task](https://app.asana.com/0/1203179076056209/1203409578249536/f) _(internal)_

## 🔍 What does this change?

- Create `GridEditorWrapper` to reuse some styles for grid editor wrappers
- Refactor value editors for different data types. Now we have `xyz-input.tsx` files instead of `xyz-editor.tsx` files, which are plain input components. So we can use those plain input components on both array and non-array editors.
- Value cell says `No values` instead of `No value` if property is an array
- Create `DraftRow` component, which contains the logic for `adding an item to the list`.
- Make `SortableRow` more generic to use it inside `DraftRow` component
- Split `value-cell-editor` to `array-editor` & `single-value-editor` to seperately handle array & single value logics
- Create some utility functions related with `EditorType`

## 🐾 Next steps

- Update boolean editor/value design
- Support multiple expected types for non-array properties

## ❓ How to test this?

1.  Checkout the branch / view the deployment
1.  Open entity editor with an array type with multiple expected types
2. Confirm that functionality works as expected

## 📹 Demo

https://user-images.githubusercontent.com/39553853/208060698-8553c52e-8a3a-4aa8-b175-3e48f2d3e151.mp4
